npeps/npep-122.md

```diff
@@ -158,7 +167,403 @@ Fourth, the ANP subject allows using pod selectors, while tenancy use cases only
+    - action: Deny
+      from:
+        - namespaces: {}
+  egress:
+    - action: Deny
+      to:
+        - namespaces: {}
+```
+BANP can also be replaced with deny-all BANP
+```yaml
+kind: BaselineAdminNetworkPolicy
+spec:
+  subject:
+    namespaces: {}
+  ingress:
+    - action: Deny
+      from:
+        - namespaces: {}
+  egress:
+    - action: Deny
+      to:
+        - namespaces: {}
+```
+
+* 4.2 strict isolation
+```yaml
+kind: TenancyNetworkPolicy
+spec:
+  tenancyLabels:
+    - "user"
+  precedence: ANP
+  action: DenyNotSameTenant
+```
+OR (second option may be more useful if there is a deny ANP in the cluster)
+```yaml
+kind: TenancyNetworkPolicy
+spec:
+  tenancyLabels:
+    - "user"
+  precedence: ANP
+  action: PassSameTenant
+---
+kind: AdminNetworkPolicy
+spec:
+  priority: 1
+  subject:
+    namespaces:
+      matchExpression:
+        - key: user
+          operator: Exists
+  ingress:
+    - action: Deny
+      from:
+        - namespaces: {}
+  egress:
+    - action: Deny
+      to:
+        - namespaces: {}
+```
+
+* 4.3 Allow internal connections for tenants
+BANP-level
+```yaml
+kind: TenancyNetworkPolicy
+spec:
+  tenancyLabels:
+    - "user"
+  precedence: BANP
+  action: PassSameTenant
+---
+kind: BaselineAdminNetworkPolicy
+spec:
+  subject:
+    namespaces:
+      matchExpression:
+        - key: user
+          operator: Exists
+  ingress:
+    - action: Deny
+      from:
+        - namespaces: {}
+  egress:
+    - action: Deny
+      to:
+        - namespaces: {}
+```
+
+ANP-level
+```yaml
+kind: TenancyNetworkPolicy
+spec:
+  tenancyLabels:
+    - "user"
+  precedence: ANP
+  action: PassSameTenant
+---
+kind: AdminNetworkPolicy
+spec:
+  priority: 1
+  subject:
+    namespaces:
+      matchExpression:
+        - key: user
+          operator: Exists
+  ingress:
+    - action: Deny
+      from:
+        - namespaces: {}
+  egress:
+    - action: Deny
+      to:
+        - namespaces: {}
+```
+
+* 4.4 Tenants interaction with (B)ANP
+* 4.4.1 allow from monitoring + deny from not same tenant
+```yaml
+kind: TenancyNetworkPolicy
+spec:
+  tenancyLabels:
+    - "user"
+  precedence: ANP
+  action: PassSameTenant
+---
+kind: AdminNetworkPolicy
+spec:
+  priority: 1
+  subject:
+    namespaces:
+      matchExpression:
+        - key: user
+          operator: Exists
+  ingress:
+    - action: Allow
+      from:
+        - namespaces:
+            namespaceSelector:
+              matchLabels:
+                kubernetes.io/metadata.name: monitoring-ns
+    - action: Deny
+      from:
+        - namespaces: {}
+```
+* 4.4.2 allow from same tenant + BANP deny all
+```yaml
+kind: TenancyNetworkPolicy
+spec:
+  tenancyLabels:
+    - "user"
+  precedence: BANP
+  action: PassSameTenant
+---
+kind: BaselineAdminNetworkPolicy
+spec:
+  subject:
+    namespaces:
+      matchExpression:
+        - key: user
+          operator: Exists
+  ingress:
+    - action: Deny
+      from:
+        - namespaces: {}
+```
+</details>
```

Comment:
all the examples provided here are priceless and well thought out, thank you Nadia!!
I have a silly question:
whichever way we decide to do, If I do:
 tenancyLabels:
    - "user" (nsA:prod, nsB:staging, nsC:prod, nsD: staging, nsE:prod)
    - "color" (nsA:red, nsB:red, nsC:blue, nsD:blue, nsE:red)
   
So subject is all namespaces that have either of these tenancyLabels set right ? (not ALL?).

then the concept of `sameTenant` means any other tenant that has `Either of these 3 keys with the same value` OR `any tenant that has ALL of these 3 keys with the same value`; is it && or || ?

so who is sameTenant here for nsA? is it nsC, nsB and nsE? or only nsE ?

Reply:
We have mostly discussed this for one label case, but I think it applies to multiple labels too (will add a section for that)
So for one label, we decided that only namespaces that have that label will be selected to avoid potential None tenant. Therefore for multiple labels, only namespaces that have all tenancy labels will be selected for the same reason. So sameTenant for nsA is namespace with user=prod AND color=red.